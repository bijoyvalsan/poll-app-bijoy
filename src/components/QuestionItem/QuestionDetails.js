import React from "react";
import PropTypes from "prop-types";

import SaveVote from "../Button";

import {
  StyledContainer,
  StyledContainerItem,
  StyledDeailsContainer,
  StyledDeailsColumn,
  StyledDeailsRow,
  StyledSubTitle,
  StyledBox
} from "./QuestionDetails.styled";

const QuestionDetails = ({ question, onVote }) => (
  <StyledBox>
    <StyledSubTitle>Question: {question.question}</StyledSubTitle>
    <StyledContainer>
      {question.choices.map((c, i) => (
        <StyledContainerItem key={c.url}>
          <StyledDeailsContainer>
            <StyledDeailsColumn>
              <StyledDeailsRow>
                Choice {i + 1}: {c.choice}
              </StyledDeailsRow>
            </StyledDeailsColumn>
            <StyledDeailsColumn>
              <StyledDeailsRow>Votes: {c.votes}</StyledDeailsRow>
            </StyledDeailsColumn>
            <StyledDeailsColumn>
              <StyledDeailsRow>Percentage: {c.percentage}%</StyledDeailsRow>
            </StyledDeailsColumn>
            <StyledDeailsColumn>
              <SaveVote.Button
                data-testid={`save-vote-${i}`}
                onClick={onVote(c.url)}
              >
                Save Vote
              </SaveVote.Button>
            </StyledDeailsColumn>
          </StyledDeailsContainer>
        </StyledContainerItem>
      ))}
    </StyledContainer>
  </StyledBox>
);

QuestionDetails.defaultProps = {
  onVote: () => {}
};

QuestionDetails.propTypes = {
  onVote: PropTypes.func,
  question: PropTypes.shape({
    id: PropTypes.string,
    question: PropTypes.string,
    choices: PropTypes.arrayOf(
      PropTypes.shape({
        choice: PropTypes.string.isRequired,
        url: PropTypes.string.isRequired,
        votes: PropTypes.number.isRequired,
        percentage: PropTypes.number.isRequired
      })
    )
  })
};

export default QuestionDetails;
