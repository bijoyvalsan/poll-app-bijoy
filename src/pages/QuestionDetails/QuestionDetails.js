import React, { useEffect } from "react";
import { useParams } from "react-router-dom";
import PropTypes from "prop-types";
import { Question, Page } from "../../components";

const QuestionDetails = ({
  getQuestionById,
  isLoading,
  question,
  saveVoteForQuestion
}) => {
  const { questionId } = useParams();
  useEffect(() => {
    getQuestionById(decodeURIComponent(questionId));
  }, [getQuestionById, questionId]);

  const handleOnVoteClick = urlPath => {
    return () => {
      saveVoteForQuestion(urlPath);
    };
  };

  return (
    <Page title="Question Details">
      <div>
        {isLoading && <Question.QuestionDetailsSkeleton />}
        {!isLoading && (
          <>
            <Question.QuestionDetails
              question={question}
              onVote={handleOnVoteClick}
            />
          </>
        )}
      </div>
    </Page>
  );
};

QuestionDetails.defaultProps = {
  getQuestionById: () => {},
  saveVoteForQuestion: () => {}
};

QuestionDetails.propTypes = {
  getQuestionById: PropTypes.func,
  saveVoteForQuestion: PropTypes.func,
  question: PropTypes.shape({
    id: PropTypes.string,
    question: PropTypes.string,
    choices: PropTypes.arrayOf(
      PropTypes.shape({
        choice: PropTypes.string.isRequired,
        url: PropTypes.string.isRequired,
        votes: PropTypes.number.isRequired,
        percentage: PropTypes.number.isRequired
      })
    )
  }),
  isLoading: PropTypes.bool.isRequired
};

export default QuestionDetails;
