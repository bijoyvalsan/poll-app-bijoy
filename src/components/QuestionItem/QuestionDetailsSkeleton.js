import React from "react";

import { StyledLoader } from "./QuestionItem.styled";
import {
  StyledDeailsContainer,
  StyledDeailsColumn,
  StyledDeailsRow,
  StyledBox,
  StyledContainer,
  StyledContainerItem
} from "./QuestionDetails.styled";

const QuestionDetailsSkeleton = () => (
  <StyledBox>
    <div>
      <StyledLoader width="100px" />
    </div>
    <StyledContainer>
      <StyledContainerItem>
        <StyledDeailsContainer>
          <StyledDeailsColumn>
            <StyledDeailsRow>
              <StyledLoader translucent width="100px" />
            </StyledDeailsRow>
          </StyledDeailsColumn>
          <StyledDeailsColumn>
            <StyledDeailsRow>
              <StyledLoader translucent width="100px" />
            </StyledDeailsRow>
          </StyledDeailsColumn>
          <StyledDeailsColumn>
            <StyledDeailsRow>
              <StyledLoader translucent width="100px" />
            </StyledDeailsRow>
          </StyledDeailsColumn>
        </StyledDeailsContainer>
      </StyledContainerItem>
      <StyledContainerItem>
        <StyledDeailsContainer>
          <StyledDeailsColumn>
            <StyledDeailsRow>
              <StyledLoader translucent width="100px" />
            </StyledDeailsRow>
          </StyledDeailsColumn>
          <StyledDeailsColumn>
            <StyledDeailsRow>
              <StyledLoader translucent width="100px" />
            </StyledDeailsRow>
          </StyledDeailsColumn>
          <StyledDeailsColumn>
            <StyledDeailsRow>
              <StyledLoader translucent width="100px" />
            </StyledDeailsRow>
          </StyledDeailsColumn>
        </StyledDeailsContainer>
      </StyledContainerItem>
      <StyledContainerItem>
        <StyledDeailsContainer>
          <StyledDeailsColumn>
            <StyledDeailsRow>
              <StyledLoader translucent width="100px" />
            </StyledDeailsRow>
          </StyledDeailsColumn>
          <StyledDeailsColumn>
            <StyledDeailsRow>
              <StyledLoader translucent width="100px" />
            </StyledDeailsRow>
          </StyledDeailsColumn>
          <StyledDeailsColumn>
            <StyledDeailsRow>
              <StyledLoader translucent width="100px" />
            </StyledDeailsRow>
          </StyledDeailsColumn>
        </StyledDeailsContainer>
      </StyledContainerItem>
    </StyledContainer>
  </StyledBox>
);

export default QuestionDetailsSkeleton;
