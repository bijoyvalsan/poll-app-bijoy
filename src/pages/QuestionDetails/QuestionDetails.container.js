import { connect } from "react-redux";

import QuestionDetails from "./QuestionDetails";

import {
  getQuestionById,
  saveVoteForQuestion
} from "../../state/actions/question";
import {
  getQuestionDetailForListPage,
  isQuestionDetailPageLoading
} from "../../state/selectors/question";

const mapStateToProps = state => ({
  question: getQuestionDetailForListPage(state),
  isLoading: isQuestionDetailPageLoading(state)
});

const mapDispatchToProps = {
  getQuestionById,
  saveVoteForQuestion
};

export default connect(mapStateToProps, mapDispatchToProps)(QuestionDetails);
