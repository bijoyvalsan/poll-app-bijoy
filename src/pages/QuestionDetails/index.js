import QuestionDetails from "./QuestionDetails.container";

export default QuestionDetails;
