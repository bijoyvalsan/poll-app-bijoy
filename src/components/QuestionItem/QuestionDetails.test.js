import React from "react";
import { render } from "@testing-library/react";

import QuestionDetails from "./QuestionDetails";
import QuestionDetailsSkeleton from "./QuestionDetailsSkeleton";

describe("QuestionDetails Component", () => {
  it("should render with default QuestionDetails styles", () => {
    const mockData = {
      id: "1",
      question: "Who am i ?",
      choices: [
        {
          choice: "a",
          url: "/test/1",
          votes: 0,
          percentage: 0
        },
        {
          choice: "b",
          url: "/test/2",
          votes: 1,
          percentage: 100
        }
      ]
    };
    const { asFragment } = render(<QuestionDetails question={mockData} />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("QuestionDetailsSkeleton Component", () => {
  it("should render with default QuestionDetailsSkeleton styles", () => {
    const { asFragment } = render(<QuestionDetailsSkeleton />);
    expect(asFragment()).toMatchSnapshot();
  });
});
