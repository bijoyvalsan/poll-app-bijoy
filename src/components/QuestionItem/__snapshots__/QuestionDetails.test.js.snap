// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionDetails Component should render with default QuestionDetails styles 1`] = `
<DocumentFragment>
  <div
    class="sc-iCoGMd dYXbjY"
  >
    <p
      class="sc-gKAaRy cOyPYX"
    >
      Question: Who am i ?
    </p>
    <div
      class="sc-gtsrHT jpJA-dN"
    >
      <div
        class="sc-dlnjwi eGFgcp"
      >
        <div
          class="sc-hKFxyN zWQlH"
        >
          <div
            class="sc-eCApnc jXgdNG"
          >
            <div
              class="sc-jSFjdj hXDXkG"
            >
              Choice 1: a
            </div>
          </div>
          <div
            class="sc-eCApnc jXgdNG"
          >
            <div
              class="sc-jSFjdj hXDXkG"
            >
              Votes: 0
            </div>
          </div>
          <div
            class="sc-eCApnc jXgdNG"
          >
            <div
              class="sc-jSFjdj hXDXkG"
            >
              Percentage: 0%
            </div>
          </div>
          <div
            class="sc-eCApnc jXgdNG"
          >
            <button
              class="sc-bdnxRM kWuiVc"
              data-testid="save-vote-0"
            >
              Save Vote
            </button>
          </div>
        </div>
      </div>
      <div
        class="sc-dlnjwi eGFgcp"
      >
        <div
          class="sc-hKFxyN zWQlH"
        >
          <div
            class="sc-eCApnc jXgdNG"
          >
            <div
              class="sc-jSFjdj hXDXkG"
            >
              Choice 2: b
            </div>
          </div>
          <div
            class="sc-eCApnc jXgdNG"
          >
            <div
              class="sc-jSFjdj hXDXkG"
            >
              Votes: 1
            </div>
          </div>
          <div
            class="sc-eCApnc jXgdNG"
          >
            <div
              class="sc-jSFjdj hXDXkG"
            >
              Percentage: 100%
            </div>
          </div>
          <div
            class="sc-eCApnc jXgdNG"
          >
            <button
              class="sc-bdnxRM kWuiVc"
              data-testid="save-vote-1"
            >
              Save Vote
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`QuestionDetailsSkeleton Component should render with default QuestionDetailsSkeleton styles 1`] = `
<DocumentFragment>
  <div
    class="sc-iCoGMd dYXbjY"
  >
    <div>
      <div
        class="sc-pNWdM sc-jrsJWt geUNDW ctAWuQ"
        width="100px"
      />
    </div>
    <div
      class="sc-gtsrHT jpJA-dN"
    >
      <div
        class="sc-dlnjwi eGFgcp"
      >
        <div
          class="sc-hKFxyN zWQlH"
        >
          <div
            class="sc-eCApnc jXgdNG"
          >
            <div
              class="sc-jSFjdj hXDXkG"
            >
              <div
                class="sc-pNWdM sc-jrsJWt cLVGPW ctAWuQ"
                width="100px"
              />
            </div>
          </div>
          <div
            class="sc-eCApnc jXgdNG"
          >
            <div
              class="sc-jSFjdj hXDXkG"
            >
              <div
                class="sc-pNWdM sc-jrsJWt cLVGPW ctAWuQ"
                width="100px"
              />
            </div>
          </div>
          <div
            class="sc-eCApnc jXgdNG"
          >
            <div
              class="sc-jSFjdj hXDXkG"
            >
              <div
                class="sc-pNWdM sc-jrsJWt cLVGPW ctAWuQ"
                width="100px"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="sc-dlnjwi eGFgcp"
      >
        <div
          class="sc-hKFxyN zWQlH"
        >
          <div
            class="sc-eCApnc jXgdNG"
          >
            <div
              class="sc-jSFjdj hXDXkG"
            >
              <div
                class="sc-pNWdM sc-jrsJWt cLVGPW ctAWuQ"
                width="100px"
              />
            </div>
          </div>
          <div
            class="sc-eCApnc jXgdNG"
          >
            <div
              class="sc-jSFjdj hXDXkG"
            >
              <div
                class="sc-pNWdM sc-jrsJWt cLVGPW ctAWuQ"
                width="100px"
              />
            </div>
          </div>
          <div
            class="sc-eCApnc jXgdNG"
          >
            <div
              class="sc-jSFjdj hXDXkG"
            >
              <div
                class="sc-pNWdM sc-jrsJWt cLVGPW ctAWuQ"
                width="100px"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="sc-dlnjwi eGFgcp"
      >
        <div
          class="sc-hKFxyN zWQlH"
        >
          <div
            class="sc-eCApnc jXgdNG"
          >
            <div
              class="sc-jSFjdj hXDXkG"
            >
              <div
                class="sc-pNWdM sc-jrsJWt cLVGPW ctAWuQ"
                width="100px"
              />
            </div>
          </div>
          <div
            class="sc-eCApnc jXgdNG"
          >
            <div
              class="sc-jSFjdj hXDXkG"
            >
              <div
                class="sc-pNWdM sc-jrsJWt cLVGPW ctAWuQ"
                width="100px"
              />
            </div>
          </div>
          <div
            class="sc-eCApnc jXgdNG"
          >
            <div
              class="sc-jSFjdj hXDXkG"
            >
              <div
                class="sc-pNWdM sc-jrsJWt cLVGPW ctAWuQ"
                width="100px"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
